### PR TITLE
React-ify beatmap discussion history page

### PR DESCRIPTION
**Pull request by nekodex**
Uses similar logic to the modding profile page to structure discussion data in such a way that they can be rendered by the beatmap-discussion react components - this way we will no longer have two distinct code paths to maintain for rendering beatmap discussions (in javascript and php...).

This also combines the users-specific discussion history page (previously at `/users/<id>/modding/discussions`) with the main index because they were functionally identical anyway.

Not sure if it's worth adding a redirect for it?

- [x] depends on #5307

### CODE REVIEW

**Comment by nanaya** [comment]
Merge conflicts :pensive:

File: resources/assets/coffee/react/beatmap-discussions-history/main.coffee [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+
+  componentWillUnmount: =>
+    $.unsubscribe '.discussionHistory'
+    $(window).off '.discussionHistory'
+
+    $(window).stop()
+
+
+  discussionUpdate: (_e, options) =>
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: looks like sometimes editing someone else's post doesn't show correct "edited by" line

File: resources/assets/coffee/react/beatmap-discussions-history/main.coffee [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+###
+
+import { Discussion } from '../beatmap-discussions/discussion'
+import { BeatmapsContext } from 'beatmap-discussions/beatmaps-context'
+import { DiscussionsContext } from 'beatmap-discussions/discussions-context'
+import * as React from 'react'
+import { a, div, img } from 'react-dom-factories'
+el = React.createElement
+
+export class Main extends React.PureComponent
```
[the remark this answers is not in the document]
Comment (nekodex) [reply resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: Most of this code is reused from `modding-profile` (which I didn't feel like porting to TypeScript)

File: resources/assets/coffee/react/beatmap-discussions-history/main.coffee (hunk 1 above)
Comment (nekodex) [reply resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: any details on how to reproduce?

Comment (nanaya) [reply resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: I think it's when you're editing others' post and there's no other post/edit by you on the page when editing it.

Comment (nekodex) [reply resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: Hmm, that shouldn't matter though, 'cuz `discussionUpdate` iterates over the related_users and adds them... weird.

File: app/Http/Controllers/BeatmapDiscussionsController.php [hunk 3]
```diff
@@ -92,13 +93,16 @@ public function index()
...
-            'user',
+            'beatmap',
+            'beatmapDiscussionVotes',
             'beatmapset',
             'startingPost',
         ])->limit($search['params']['limit'] + 1);
 
-        $discussions = new Paginator(
-            $query->get(),
+        $discussions = $query->get();
```
Comment (nanaya) [inline app/Http/Controllers/BeatmapDiscussionsController.php]: This contains one extra item which is used to tell `Paginator` whether or not there are more pages. The last item shouldn't be actually rendered.

File: resources/assets/coffee/react/beatmap-discussions-history/main.coffee [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  render: =>
+    el DiscussionsContext.Provider,
+      value: @discussions()
+      el BeatmapsContext.Provider,
+        value: @beatmaps()
+        div className: 'modding-profile-list modding-profile-list--index',
+          if @props.discussions.length == 0
+            div className: 'modding-profile-list__empty', osu.trans('beatmap_discussions.index.none_found')
+          else
+            [
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: the array does nothing

File: resources/assets/coffee/react/beatmap-discussions-history/main.coffee [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    # - not privileged (deleted discussion)
+    # - deleted beatmap
+    @cache.discussions ?= _ @state.relatedDiscussions
+                            .filter (d) -> !_.isEmpty(d)
+                            .keyBy 'id'
+                            .value()
+
+
+  beatmaps: =>
+    @cache.beatmaps if @cache.beatmaps?
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: I think this needs `return` keyword.

File: resources/assets/coffee/react/beatmap-discussions-history/main.coffee [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    beatmaps = _.map(@discussions(), (d) => d.beatmap)
+                .filter((b) => b != undefined)
+
+    @cache.beatmaps ?= _.keyBy(beatmaps, 'id')
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmap-discussions-history/main.coffee]: Can use plain `=`

**Pull request merged**